### PR TITLE
Add Context to RequestOptions

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR adds `Context` to `RequestOptions`. Having `Context` on `RequestOptions` allows for a simplification of LLC APIs. `RequestOptions` should only be used in LLC APIs, so this won't affect current SDKs.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java [hunk 1]
```diff
@@ -121,20 +122,28 @@ public Object invoke(Object proxy, final Method method, Object[] args) {
         try {
             final SwaggerMethodParser methodParser = getMethodParser(method);
             final HttpRequest request = createHttpRequest(methodParser, args);
-            Context context = methodParser.setContext(args)
-                .addData("caller-method", methodParser.getFullyQualifiedMethodName())
+            Context context = methodParser.setContext(args);
+
+            RequestOptions options = methodParser.setRequestOptions(args);
+            if (options != null) {
+                options.getRequestCallback().accept(request);
```
Comment (weidongxu-microsoft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:129]: This `accept` is now before `request.setBody(validateLength(request));`, will it be a problem?

Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:129]: Hmm, that's a good question. For safety, given how late this change is going in, I'll have two if blocks to use the `RequestOptions`.

**Review by lmolkova — APPROVED** [review]
Can you please add tests?

**Review by pallavit — APPROVED** [review]
:shipit:

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java [hunk 2]
```diff
@@ -121,16 +122,31 @@ public Object invoke(Object proxy, final Method method, Object[] args) {
...
+            context = context.addData("caller-method", methodParser.getFullyQualifiedMethodName())
                 .addData("azure-eagerly-read-response", shouldEagerlyReadResponse(methodParser.getReturnType()));
             context = startTracingSpan(method, context);
 
             if (request.getBody() != null) {
                 request.setBody(validateLength(request));
             }
 
-            RequestOptions options = methodParser.setRequestOptions(args);
+            // This second if check on 'options' is to maintain previous execution ordering, as the change to include
```
Comment (joshfree) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:148]: can you rephrase this?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java [hunk 3]
```diff
@@ -121,16 +122,31 @@ public Object invoke(Object proxy, final Method method, Object[] args) {
...
+            // 'Context' in 'RequestOptions' came in late in a development cycle.
```
Comment (joshfree) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:149]: should there be a tracking issue referenced in the comment?

File: sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java [hunk 4]
```diff
@@ -95,12 +95,14 @@ private Context(Context parent, Object key, Object value) {
      *
      * <!-- src_embed com.azure.core.util.context.addData#object-object -->
      * <pre>
-     * &#47;&#47; Users can send parent span information and pass additional metadata to attach to spans of the calling methods
-     * &#47;&#47; using the Context object
+     * &#47;&#47; Users can pass parent trace context information and additional metadata to attach to spans create by SDKs
```
Comment (joshfree) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java:98]: s/create/created/

File: sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java [hunk 5]
```diff
@@ -95,12 +95,14 @@ private Context(Context parent, Object key, Object value) {
...
+     * &#47;&#47; using the com.azure.core.util.Context object.
      * final String hostNameValue = &quot;host-name-value&quot;;
      * final String entityPathValue = &quot;entity-path-value&quot;;
-     * final String userParentSpan = &quot;user-parent-span&quot;;
-     * Context parentSpanContext = new Context&#40;PARENT_TRACE_CONTEXT_KEY, userParentSpan&#41;;
+     *
+     * &#47;&#47; TraceContext represents tracing solution context type - io.opentelemetry.context.Context for for OpenTelemetry.
```
Comment (joshfree) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java:103]: double "for"

Comment (joshfree) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java:103]: missing 'a' -- "TraceContext represents a tracing solution context type"

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java [hunk 6]
```diff
@@ -121,16 +122,31 @@ public Object invoke(Object proxy, final Method method, Object[] args) {
...
             final SwaggerMethodParser methodParser = getMethodParser(method);
             final HttpRequest request = createHttpRequest(methodParser, args);
-            Context context = methodParser.setContext(args)
-                .addData("caller-method", methodParser.getFullyQualifiedMethodName())
+            Context context = methodParser.setContext(args);
+
+            RequestOptions options = methodParser.setRequestOptions(args);
+            if (options != null) {
+                Context optionsContext = options.getContext();
+                if (optionsContext != null && optionsContext != Context.NONE) {
```
Comment (joshfree) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:130]: should we add a perf test for this?
/cc @g2vinay

**Review by joshfree — APPROVED** [review]
